astropy/io/fits/hdu/image.py

```diff
@@ -246,7 +246,7 @@ def data(self):
 
     @data.setter
     def data(self, data):
-        if 'data' in self.__dict__:
+        if 'data' in self.__dict__ and self.__dict__['data'] is not None:
```

Comment:
Not sure it makes a difference, but I wonder if the test for `not None` should be just before `was_unsigned` to make sure that `_data_replaced` is set even if `data` was `None`.